Add true and false as PseudoTypes

mvriel opened the pull request:
In this change, I have made more allowance for identifying types as
PseudoTypes and to ensure they are identified as both their underlying
and pseudo-type.

Using instanceof, you can identify whether it is the PseudoType, a
pseudotype, but also as the underlying type.

mvriel committed:
Add true and false as PseudoTypes

In this change, I have made more allowance for identifying types as
PseudoTypes and to ensure they are identified as both their underlying
and pseudo-type.

Using instanceof, you can identify whether it is the PseudoType, a
pseudotype, but also as the underlying type.

mvriel committed:
Correct issues in tests

mvriel committed:
Position return type : correctly

mvriel committed:
Fix linting issues

mvriel committed:
Merge branch 'master' of github.com:phpDocumentor/TypeResolver into add-pseudo-type-support-and-use-true-and-false

mvriel committed:
Added immutable annotations

mvriel committed:
Add test for Boolean type too

mvriel committed:
Merge branch '1.x' of github.com:phpDocumentor/TypeResolver into add-pseudo-type-support-and-use-true-and-false

mvriel committed:
Retarget to 1.x and fix conflicts

mvriel committed:
Fix linting and pre-commit test

jaapio commented on src/PseudoTypes/False_.php:
```diff
@@ -0,0 +1,36 @@
+ *
+ * @psalm-immutable
+ */
+final class False_ extends Boolean implements PseudoType
```
Moving this class to another namespace isn't possible :-( becaus of BC.

mvriel replied:
As discussed on Slack: I'll add a class_alias to the old name. The new namespace helps clarify

jaapio replied:
It might be good to introduce a deprecation warning on those aliases :-)

mvriel replied:
I have been unable to add a deprecation warning; at least: I don't know how I would accomplish that

jaapio commented on src/PseudoTypes/True_.php:
```diff
@@ -0,0 +1,36 @@
+ *
+ * @psalm-immutable
+ */
+final class True_ extends Boolean implements PseudoType
```
Moving this class to another namespace isn't possible :-( becaus of BC.

mvriel replied:
As discussed on Slack: I'll add a class_alias to the old name. The new namespace helps clarify

mvriel committed:
Support deprecated FQSENs for BC

mvriel committed:
Fix styling issues

mvriel commented:
The BC break mentioned in the build is mitigated by using class_alias in the False_ and True_ classes. The tool is apparently unable to detect that and I even added a test to check the BC

mvriel committed:
Disable phpcs sniff for BC compat

In order to keep BC compatiblity, I had to introduce class_alias calls
in the True_ and False_ classes. This has a side-effect, which the phpcs
fails on.

jaapio committed:
Add extra check for BC
